### PR TITLE
Exclude System.Func from the completion list when the user is likely …

### LINKED ISSUE #4208: VB Intellisense for predicate includes Func(of ...)
In VS <= 2013, when entering an argument of type Func(of ...) in a function call, Func(of ...) was not shown in the intellisense popup. As such, typing "fu"+tab was usually enough to insert "function". Now Func(of ...) shows in the list, which is never required, and requires an extra keystroke to move to "function".

e.g. type the below in both 2013 and 2015 and observe difference in intellisense list shown:

enumerable.range(0, 10).select(fu

### PR DESCRIPTION
**Pull request by rchande**
…to be typing Function

Fixes #4208 

cc @dpoeschl @jasonmalinowski @balajikris @davkean @DustinCampbell @Pilchie

### CODE REVIEW

File: src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationService.vb [hunk 1]
```diff
@@ -135,9 +136,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Recommendations
...
+            Dim typeInferrer = context.GetLanguageService(Of ITypeInferenceService)()
+            Dim inferredTypes = typeInferrer.InferTypes(context.SemanticModel, context.Position, cancellationToken:=cancellationToken)
+            If inferredTypes?.Any(Function(t) t IsNot Nothing AndAlso IsFunc(t.OriginalDefinition)) Then
+                symbols = symbols.Where(Function(s) Not IsFunc(TryCast(s, ITypeSymbol)))
+            End If
+
             Return symbols.Where(Function(s) FilterEventsAndGeneratedSymbols(Nothing, s))
         End Function
 
+        Private Function IsFunc(t As ITypeSymbol) As Boolean
```
Comment (jasonmalinowski) [inline src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationService.vb]: Should we just filter all delegate types in general? In general we list type names in a parameter position to allow for static method calling or use of static properties. Should Delegates be generically exempted?

Comment (jasonmalinowski) [inline src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationService.vb]: Consider: somebody is using a library that also has delegate types that starts with "Func".

File: src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationService.vb [hunk 2]
```diff
@@ -135,9 +136,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Recommendations
...
+            Return t IsNot Nothing AndAlso
+                    t.Name = "Func" AndAlso
+                    t.GetArity() > 0 AndAlso
+                    t.ContainingNamespace IsNot Nothing AndAlso
+                    t.ContainingNamespace.Name = "System"
```
Comment (jasonmalinowski) [inline src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationService.vb]: Library.System.Func, what do you want to do about that?

**Comment by rchande** [comment]
@jasonmalinowski That sounds better; I've switched to hiding all delegate types when we infer a delegate type.

**Comment by jasonmalinowski** [comment]
Is the type inference check necessary? It would seem it's either valid ever or not. Also, what about explicit calls to Delegate.CreateDelegate?

**Comment by rchande** [comment]
The type inference is necessary to differentiate typing something like ```VB Dim x As System.Func(Of Integer)``` vs typing a lambda where Func( gets in the way of typing Function.

**Comment by jasonmalinowski** [comment]
But can't we know that syntactically? One of those is a type context, one is an expression context.

**Comment by DustinCampbell** [comment]
I don't think this is right fix. In VS 2013, we include Func(Of ...) in this case if the completion list is explicitly invoked, but not if it's invoked by typing a character.

**Invoked with Ctrl+Space**:

image

**Invoked by typing**:

image

**Comment by Pilchie** [comment]
Why?  In what scenario do you ever want to type Func(Of T) there?

**Comment by DustinCampbell** [comment]
The fix is broader than just there. The code is in "GetUnqualifiedSymbolsForExpressionOrStatementContext". So, it would also stop Func(Of T) from appearing here:

```VB
Sub M()
    $$
End Sub
```

There need to be a lot more negative tests to ensure we don't break the simple cases just to make that one work.

**Comment by DustinCampbell** [comment]
I chatted with Jason, and agree that there really isn't a reason to show any delegates in these positions. However, we should be sure to still show System.Delegate itself. We also need more positive and negative unit tests to make sure we don't break anything. For example, I'd hate to see delegates stop showing up in ```GetType($$)```.

**Comment by rchande** [comment]
@jasonmalinowski @Pilchie @DustinCampbell Added lots of tests and got rid of type inference checks.

**Comment by DustinCampbell** [comment]
:+1:

File: src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationService.vb [hunk 3]
```diff
@@ -133,11 +134,30 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Recommendations
...
             Return symbols.Where(Function(s) FilterEventsAndGeneratedSymbols(Nothing, s))
         End Function
 
+        Private Function IsSystemFunc(s As ISymbol) As Boolean
+            Dim namedTypeSymbol = TryCast(s, INamedTypeSymbol)
+            Return namedTypeSymbol IsNot Nothing AndAlso
+                    namedTypeSymbol.Name = "Func" AndAlso
+                    namedTypeSymbol.GetArity() > 0 AndAlso
+                    namedTypeSymbol.ContainingNamespace IsNot Nothing AndAlso
+                    namedTypeSymbol.ContainingNamespace.Name = "System" AndAlso
```
Comment (Pilchie) [inline src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationService.vb]: You could use `?.` here.

**Comment by Pilchie** [comment]
:+1:

**Comment by jasonmalinowski** [comment]
Per @DustinCampbell and my earlier conversation, I thought we were going to go with generally unlisting delegates versus a specific hack for System.Func. Was the specific hack ultimately needed for some reason?

**Comment by DustinCampbell** [comment]
Sorry, yeah, I only checked the test coverage. Let me pull back my :+1: :smile:

**Comment by davkean** [comment]
What was the resolution to this?

**Comment by jasonmalinowski** [comment]
We reactivated the bug to revisit this.

**Pull request merged**